apiary.apib

```diff
@@ -284,58 +284,70 @@ session.
 the Queue functionality in your own application or even hook it up to a
-completely different Queue system  it can be accomplished in a few different ways:
-
-#### Set a Javascript specific endpoint
+completely different Queue system it can be accomplished in a few different ways:
 
+#### Set a custom queue endpoint
 
 You can configure the 'QUEUE_ENDPOINT' in the Javascript snippet to specify where
 we should fire a request each time a new user wants to start a session.
 
 As long as an agent is Queue'd, we will fire a POST request to this endpoint
-containing the following body:
-
+containing the following JSON in request body:
 
 
 ```
 {
-    'url': <actual url to start a session on>,
-    'id': <session_id>,
-    'viewer_link': <the link the agent needs to open to join the session>,
-    'ip_address': <ip address of the website visitor>
+    "url": <actual url to start a session on>,
+    "id": <session_id>,
+    "viewer_link": <the link the agent needs to open to join the session>,
+    "ip_address": <ip address of the website visitor>
 }
 ```
 
 Your controller that handles this request can then display or dispatch the
 viewer link to any of the available agents.
 
+Endpoint is expected to respond with a JSON object. It must contain at least one
+field - `identifier`, which should be a unique string identifying the current session:
+
+```
+{
+    "identifier": "<unique_id>",
+    ...additional session metadata...
+}
+```
+
+We will pass the endpoint response to `QUEUE_CALLBACK` every time (see below), so you can
+store some session metadata in additional fields.
+
 
 #### Set up custom Javascript Queue functionality
 
 If you want to handle the Queue functionality yourself completely you can also
-provide us with two callbacks:
+provide us with a custom Javascript function object in `QUEUE_HANDLER` option.
 
- - `QUEUE_HANDLER`
- - `QUEUE_CALLBACK`
+If it is specified, we will call `QUEUE_HANDLER` when a user requests a Surfly
+session (instead of making AJAX call to `QUEUE_ENDPOINT`). `QUEUE_HANDLER`
+will be given two arguments:
 
-We will call `QUEUE_HANDLER` a user is requesting a Surfly session. The argument
-to this call is a JSON object with the following parameters:
+- JSON object with session information (see request format for `QUEUE_ENDPOINT` above)
+- a success callback function. `QUEUE_HANDLER` must call this callback after a successful
+  queue join. You must pass a session metadata object as a first argument (see
+  response format for `QUEUE_ENDPOINT`)
 
 
- - id : session id
- - viewer_link: the link the agent should open
- - ip_address: the ip address of the client
+#### Tracking queue status
 
-If you use any one of the above Queue functionality `QUEUE_HANDLER` or `QUEUE_ENDPOINT`
-in your custom page.
+If you use `QUEUE_HANDLER` or `QUEUE_ENDPOINT`, you can also set the
+`QUEUE_CALLBACK` Javascript function to trace the status of the queue.
```

Comment:
That's a good question. Currently, if `QUEUE_HANDLER` and `QUEUE_ENDPOINT` are not set, we don't call `QUEUE_CALLBACK` in some places. I believe we should either ignore it if none of those two is defined, or make it independent (so call it even those are not set). In the latter case, it will get the response from surfly endpoint, but that's fine I guess.